core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceService.scala

```diff
@@ -20,13 +21,67 @@ import org.broadinstitute.dsde.rawls.model.{
 
 import java.util.UUID
 import scala.jdk.CollectionConverters._
+import scala.util.Try
 
 /**
   * Responsible for aggregating workspace data together from various sources (i.e, WSM).
   * @param workspaceManagerDAO DAO for talking to WSM
   */
 class AggregatedWorkspaceService(workspaceManagerDAO: WorkspaceManagerDAO) extends LazyLogging {
 
+  /**
+    * Given a list of workspaces, aggregates any available workspace information from workspace manager (WSM).
+    *
+    * Pre-fetches all available workspace data from WSM for the given context.
+    *
+    * If the WSM workspace is a legacy Rawls workspace, it is assumed to be GCP, and the googleProjectId from the
+    * provided workspace will be echo'd back out, and it will not attempt to match the workspace with WSM data
+    *
+    * Exceptions encountered when matching a multi-cloud/WSM workspace to the data from WSM are reported in the errorMessage of the workspace
+    *
+    * @param workspaces The list of source rawls workspaces
+    * @param ctx Rawls request and tracing context.
+    *
+    * @throws WorkspaceAggregationException when an error is encountered pulling data from aggregating upstream systems
+
+    */
+  def fetchAggregatedWorkspaces(workspaces: Seq[Workspace], ctx: RawlsRequestContext): Seq[AggregatedWorkspace] = {
+    val span = startSpanWithParent("listWorkspacesFromWorkspaceManager", ctx.tracingSpan.orNull)
+    try {
+      val wsmResponse = workspaceManagerDAO.listWorkspaces(ctx)
+      workspaces.map(workspace =>
+        workspace.workspaceType match {
+          case WorkspaceType.RawlsWorkspace =>
+            AggregatedWorkspace(workspace,
+                                Some(workspace.googleProjectId),
+                                azureCloudContext = None,
+                                policies = List.empty
+            )
+          case WorkspaceType.McWorkspace =>
+            val id = workspace.workspaceIdAsUUID
+            wsmResponse
+              .find(_.getId == id)
+              .map(wsmInfo =>
+                Try(aggregateMCWorkspaceWithWSMInfo(workspace, wsmInfo)).recover {
+                  case e: InvalidCloudContextException =>
+                    val ws = workspace.copy(errorMessage =
+                      Some(s"Invalid Cloud Context from Workspace Manager: ${e.getMessage}")
+                    )
+                    AggregatedWorkspace(ws, None, None, policies = List.empty)
+                }.get
+              )
+              .getOrElse {
+                val ws = workspace.copy(errorMessage = Some("Workspace not found in Workspace Manager"))
+                AggregatedWorkspace(ws, None, None, policies = List.empty)
+              }
+        }
+      )
+    } catch {
+      case e: ApiException => throw new WorkspaceAggregationException(errorReport = ErrorReport(e.getCode, e))
```

Comment:
Maybe we just shouldn't be catching anything here, since we're not querying a specific workspace, then?
403s and 404s don't really make sense as results from the list query.